wl-sim/main/WL_Flash.cpp

```diff
@@ -0,0 +1,292 @@
+uint32_t erase_count[SECTOR_COUNT + 1] = {0};
+size_t last_erase_sector = 0;
+size_t restarted = 0;
+
+// key_len is half of bit length of address space
+// so 8bit key is for (up to) 16bit address space
+// 65K sectors is plenty: up to 32MB of 512B sectors
+// or 256MB of 4K sectors
+uint8_t keys[3] = {0};
+uint8_t B = 0, MSB = 0, LSB = 0;
+uint32_t feistel_calls = 0;
+uint32_t feistel_cycle_walks = 0;
+static bool feistel = false;
+
+void init_feistel()
+{
+    feistel = true;
+
+    uint16_t sector_count = SECTOR_COUNT;
+    ESP_LOGI(TAG, "%s: N=%u", __func__, sector_count);
+    ESP_LOGI(TAG, "sizeof(size_t)=%u", sizeof(size_t));
+
+    // B = log2(sector_count)
+    for (B = 0; sector_count; B++)
+        sector_count >>= 1;
+
+    LSB = (B+1)/2;
+    MSB = B - LSB;
+    ESP_LOGI(TAG, "%s: B=%u, MSB=%u, LSB=%u", __func__, B, MSB, LSB);
+
+    uint8_t key_len = (B + 1)/2;
+
+    ESP_LOGI(TAG, "%s: key_len=%u", __func__, key_len);
+
+    for (auto i = 0; i < 3; i++)
+        keys[i] = rand() % UINT8_MAX;
+
+    ESP_LOGI(TAG, "%s: generated 8bit keys (%u, %u, %u) ", __func__, keys[0], keys[1], keys[2]);
+
+}
+
+static uint32_t feistel_function(uint32_t msb, uint32_t key)
+{
+    ESP_LOGV(TAG, "%s: msb=0x%x key=0x%x", __func__, msb, key);
+    ESP_LOGV(TAG, "%s: msb xor key = 0x%x", __func__, (msb ^ key));
+    ESP_LOGV(TAG, "%s: return (msb xor key)^2 = 0x%x", __func__, (msb ^ key)*(msb ^ key));
+    return (msb ^ key) * (msb ^ key);
+}
+
+size_t feistel_network(size_t logical_addr)
+{
+    feistel_calls++;
+    size_t addr = logical_addr;
+
+round:
+    size_t sector_addr = addr / SECTOR_SIZE;
+
+    //               |       B       |
+    //               |<-MSB->|<-LSB->|
+    // logical_addr: |  msb  |  lsb  |
+
+    size_t B_mask = ~( (~(size_t)0) << B );
+
+    size_t LSB_mask = ~( (~(size_t)0) << LSB );
+    size_t MSB_mask = ( (~(size_t)0) << LSB ) & B_mask;
+
+    size_t msb, lsb, _msb, _lsb, randomized_addr;
+
+    for (auto i = 0; i < 3; i++) {
+        msb = sector_addr >> LSB;
+        lsb = sector_addr & LSB_mask;
+
+        _msb = msb;
+        // mask output of F to also be |LSB| for XORing with lsb
+        _lsb = (lsb ^ (feistel_function(msb, keys[i]) & LSB_mask));
+
+        // swap lsb and msb
+        sector_addr = (_lsb << MSB) | _msb;
+        ESP_LOGV(TAG, "%s: msb=0x%x lsb=0x%x sector_addr=0x%x", __func__, _msb, _lsb, sector_addr);
+    }
+    randomized_addr = sector_addr;
+    ESP_LOGD(TAG, "%s: randomized_addr=0x%x", __func__, randomized_addr);
+
+    if (randomized_addr >= SECTOR_COUNT) {
+        addr = randomized_addr * SECTOR_SIZE;
+        feistel_cycle_walks++;
+#if 0
+        ESP_LOGE(TAG, "%s: logical=0x%x => randomized 0x%x outside of domain 0x%x, cycle walk", __func__, logical_addr, randomized_addr, SECTOR_COUNT);
+
+        if (feistel_cycle_walks > 20) {
+            ESP_LOGE(TAG, "%s: feistel cycle walks critical, exiting", __func__);
+            exit(1);
+        }
+#endif
+        goto round;
```

Comment:
goto? maybe do-while does the same job in more C++ way

Reply:
so far I like the `goto` there, not gonna lie. will keep if for now but I guess I'll have to change it at some point